design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,566 @@
+  the namespace where the HostClaim is must be a member of the list.
+* ``nameMatches``: the content of this field is a regular expression
+  (in Golang re2 syntax). The name of the namespace where the HostClaim is located
+  must match this regular expression.
+* ``withLabels``: the content of the field is a list of strings representing
+  labels. The namespace of the HostClaim must have a label belonging to the
+  list.
+
+To successfully match, the namespace must fulfill the constraints of all the
+defined fields. An empty HostDeployPolicy authorizes all namespaces.
+A HostClaim is authorized to bind a BareMetalHost if and only if the namespace
+of the HostClaim is successfully filtered by at least one HostDeployPolicy in
+the namespace of the BareMetalHost. In that case, the binding between the
+HostClaim and the BareMetalHost is authorized.
+
+A HostDeployPolicy specification also contains a field ``metadataCopy``. It defines
+rules for which annotations are copied (``annotationsMatch``) and which
+labels are copied (``labelsMatch``). Both fields contain regular expressions.
+
+For example,
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: HostDeployPolicy
+metadata:
+  name: policy
+  namespace: infra
+spec:
+  hostClaimNamespaces:
+    namespaces:
+    - cluster1
+    - cluster2
+  metadataCopy:
+    annotationsMatch: 'mydomain/.*'
+    labelsMatch: 'selectors/.*'
+```
+
+will make the BareMetalHosts in the namespace ``infra`` available to HostClaims
+in the namespaces ``cluster1`` and ``cluster2``. All labels in the domain ``selectors``
+will be available for selection as annotations in ``mydomain`` for constructing
+cloud-init ``metadata``.
+
+#### Multi-tenancy
+
+As an infrastructure administrator I would like to host several isolated
+clusters.
+
+All the servers in the data-center are registered as BareMetalHost in one or
+several namespaces under the control of the infrastructure manager. Namespaces
+are created for each tenant of the infrastructure. They create
+standard cluster definitions in those namespaces.
+
+When the cluster is started, a HostClaim is created for each Metal3Machine
+associated with the cluster. The ``hostSelector`` is inherited from the
+Metal3Machine. As in the original workflow, it is used to choose the BareMetalHost
+associated with the cluster, but the associated BareMetalHost is not in the same
+namespace as the HostClaim. The exact definition of the BareMetalHost remains
+hidden from the cluster user; only a small parts of its status is copied
+back to the HostClaim resource (actual power state). The status of the HostClaim
+contains a reference to the BareMetalHost and so to the HardwareData that is
+readable by the end user.
+
+With the help of HardwareData, the data template controller has enough details to
+compute the different secrets (userData, metaData and networkData) associated to
+the Metal3Machine. These secrets are linked to the HostClaim and, ultimately, to
+the BareMetalHost.
```

Comment:
I would like to clarify how the secrets are handled. Are they created in the same namespace as the HostClaim and then copied to the BareMetalHost namespace?

Reply:
Added a few lines. As usual the tricky part is garbage collection when the bmh is reused.